asllib/diet.ml

```diff
@@ -421,6 +433,23 @@ module Make (Elt : ELT) = struct
     in
     loop empty t n
 
+  let _make_intervals =
+    let rec loop acc ((x, y) as i) = function
+      | [] -> List.map (fun (a, b) -> Interval.make a b) (i :: acc)
+      | z :: t ->
+          let y' = Elt.succ y in
+          if eq y' z then loop acc (x, y') t
+          else (
+            assert (y' < z);
+            loop (i :: acc) (z, z) t)
+    in
+    function [] -> [] | x :: t -> loop [] (x, x) t
+
+  let of_list li =
+    List.sort_uniq Elt.compare li
+    |> _make_intervals
+    |> List.fold_left (fun acc elt -> add elt acc) empty
```

Comment:
Another suggestion:

1.  Get rid of an intermediate list of pairs.
2.  Trading the assert for a local function.
```suggestion
   let of_list =

    let _make_intervals =
      let rec loop acc x y = function
        | [] -> Interval.make x y :: acc
        | z :: t ->
            let y' = Elt.succ y in
            if eq y' z then loop acc x y t
            else
              loop (Interval.make x y :: acc) z z t
    in
    function [] -> [] | x :: t -> loop [] x x t in

    fun li ->
      List.sort_uniq Elt.compare li
      |> _make_intervals
      |> List.fold_left (fun acc elt -> add elt acc) empty

```